### PR TITLE
101 quick fix for persistence entity

### LINKED ISSUE #101: Add quick-fix for Jakarta Persistence @Entity diagnostic
If a class is annotated with jakarta.persistence.Entity:
1. The entity class must have a public or protected no-arg constructor (the class may have other constructors).
2. The entity class must not be final. No methods or persistent instance variables of the entity class may be final.

Add quick-fixes to this diagnostic to:
- add a public or no-arg constructor if none present
- remove the final keyword from the class, methods, or persistent instances variables if present

See https://jakarta.ee/specifications/persistence/3.0/jakarta-persistence-spec-3.0.html#a18

### PR DESCRIPTION
**Pull request by lamminade**
fixes #101 

Adds quick fixes to diagnostics to:
- add a public or no-arg constructor if none are present and a constructor with arguments is present
- remove the final keyword from the class, methods, or persistent instances variables if present

Adding constructor:
Screen Shot 2021-03-05 at 12 32 26 PM

Removing 'final' from class:
Screen Shot 2021-03-05 at 1 27 30 PM

Removing 'final' from variables:
Screen Shot 2021-03-05 at 1 27 43 PM
Screen Shot 2021-03-05 at 1 28 21 PM

Removing 'final' from methods:
Screen Shot 2021-03-05 at 1 28 30 PM

Result:
Screen Shot 2021-03-05 at 1 29 09 PM

### CODE REVIEW

File: jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/codeAction/proposal/ModifyModifiersProposal.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+     */
+    public ModifyModifiersProposal(String label, ICompilationUnit targetCU, CompilationUnit invocationNode,
+            IBinding binding, int relevance, ASTNode coveredNode, String visibility) {
+        super(label, CodeActionKind.QuickFix, targetCU, null, relevance);
+        this.invocationNode = invocationNode;
+        this.binding = binding;
+        this.coveredNode = coveredNode;
+        this.modifiersToAdd = Arrays.asList(visibility);
+        this.modifiersToRemove = new ArrayList<>();
+    }
```
Comment (kathrynkodama) [inline jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/codeAction/proposal/ModifyModifiersProposal.java:77]: If this constructor is not being used we can remove it

File: jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/codeAction/proposal/ModifyModifiersProposal.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+    public ModifyModifiersProposal(String label, ICompilationUnit targetCU, CompilationUnit invocationNode,
```
Comment (kathrynkodama) [inline jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/codeAction/proposal/ModifyModifiersProposal.java:79]: Let's add a comment to this method to clarify that it accepts both a list of modifiers to add and a list of modifiers to remove. 

Add a similar comment to the below method to clarify that it only accepts a list of modifiers to add

File: jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/jdt/persistence/PersistenceEntityQuickFix.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        List<CodeAction> codeActions = new ArrayList<>();
+
+        // option for protected constructor
+        String name = "Add a no-arg protected constructor to this class";
+        ChangeCorrectionProposal proposal = new AddConstructorProposal(name,
+                context.getCompilationUnit(), context.getASTRoot(), parentType, 0);
+        CodeAction codeAction = context.convertToCodeAction(proposal, diagnostic);
+
+        if (codeAction != null) {
+            codeAction.setTitle(name);
```
Comment (kathrynkodama) [inline jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/jdt/persistence/PersistenceEntityQuickFix.java:107]: `setTitle` is handled int he `convertToCodeaAction` method so we can remove this and the following `setTitle()`

Comment (lamminade) [reply jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/jdt/persistence/PersistenceEntityQuickFix.java]: 56c1b47

File: jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/codeAction/proposal/ModifyModifiersProposal.java (hunk 2 above)
Comment (lamminade) [reply jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/codeAction/proposal/ModifyModifiersProposal.java]: 56c1b47

File: jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/codeAction/proposal/ModifyModifiersProposal.java (hunk 1 above)
Comment (lamminade) [reply jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/codeAction/proposal/ModifyModifiersProposal.java]: 56c1b47

File: jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/jdt/persistence/PersistenceEntityQuickFix.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        }
+        
+        String name = "Remove the 'final' modifier from this ";
+        name = name.concat(type);
+        ChangeCorrectionProposal proposal = new ModifyModifiersProposal(name, context.getCompilationUnit(), 
+                context.getASTRoot(), parentType, 0, coveredNode, new ArrayList<>(), Arrays.asList("final"));
+        CodeAction codeAction = context.convertToCodeAction(proposal, diagnostic);
+        
+        if (codeAction != null) {
+            codeAction.setTitle(name);
```
Comment (lamminade) [inline jakarta-eclipse/org.eclipse.lsp4jakarta.core/src/org/jakarta/jdt/persistence/PersistenceEntityQuickFix.java:144]:
```suggestion
```

**Review by kathrynkodama — APPROVED** [review]
Looks good, thanks @lamminade

**Pull request merged**